viewer/udSample.h

```diff
@@ -18,27 +24,26 @@ struct udSampleRenderInfo
   struct SDL_Texture *pSDLTexture;
 };
 
-typedef void (udSampleInit)(void **ppSampleData, const udSampleRenderInfo &info);
-typedef void (udSampleDeinit)(void *pSampleData);
-typedef void (udSampleRender)(void *pSampleData, const udSampleRenderInfo &info);
 
-struct udSample
+// Base class for samples
+class udSample
 {
-  const char *pName;
-
-  udSampleInit *pInit; // Called when the Sample is 'activated'
-  udSampleDeinit *pDeinit; // Called when the Sample is 'deactivated'
-  udSampleRender *pRender; // Called each frame
+public:
+  udSample() { samples.push_back(this); }
+  virtual ~udSample() {}
+
+  virtual const char *GetName() const = 0;                          // Return the name of the sample
+  virtual udError Init(udSampleRenderInfo &info) = 0;               // Initialise for running the sample
+  virtual udError Deinit() = 0;                                     // Tear-down resources to run another sample, or to re-run from the start
+  virtual udError Render(udSampleRenderInfo &info) = 0;             // Render one frame
+  virtual udError Event(udSampleRenderInfo &, const SDL_Event &);   // Handle SDL events, return udE_Success if handled, or udE_NothingToDo if not handled, all other returns are errors
```

Comment:
Similar comment as above, `bool` is probably more than enough. Despite the "all other returns are errors", it's treated as a bool with `udE_NothingToDo` being the only value checked against.